arch/risc-v/src/k230/k230_rptun.c

```diff
@@ -0,0 +1,501 @@
+
+  return 0;
+}
+
+#ifdef CONFIG_K230_RPTUN_MASTER
+/****************************************************************************
+ * Name: rp_reset
+ ****************************************************************************/
+
+static void rp_reset(struct rptun_dev_s *dev, int value)
+{
+  /* Only do hard reset for big core */
+
+  if (k230_hart_is_big())
+    {
+      rperr("Can't reset little core!\n");
+    }
+  else
+    {
+      k230_hart_big_stop();
+      k230_hart_big_boot(0);
+    }
+}
+
+/****************************************************************************
+ * Name: rp_panic
+ ****************************************************************************/
+
+static void rp_panic(struct rptun_dev_s *dev)
+{
+#ifdef CONFIG_K230_RPTUN_IPI
+  k230_ipi_signal(RPTUN_IPI_REMOTE_PANIC);
+#else
+  rperr("TODO: ask remote to panic.\n");
+#endif
+}
+#endif
+
+#ifndef CONFIG_K230_RPTUN_IPI
+
+static void k230_rptun_work(void *arg)
+{
+  struct k230_rptun_dev_s *dev = arg;
+  bool should_notify = false;
+
+  if (dev->shmem != NULL)
+    {
+      if (dev->master && dev->seq != dev->shmem->seq_mst)
+        {
+          dev->seq = dev->shmem->seq_mst;
+          should_notify = true;
+        }
+      else if (!dev->master && dev->seq != dev->shmem->seq_rmt)
+        {
+          dev->seq = dev->shmem->seq_rmt;
+          should_notify = true;
+        }
+
+      if (should_notify && dev->callback != NULL)
+        {
+          dev->callback(dev->arg, RPTUN_NOTIFY_ALL);
+        }
+    }
+
+  work_queue(HPWORK, &dev->work, k230_rptun_work, dev,
+             CONFIG_K230_RPTUN_WORK_DELAY);
+}
+
+#endif /* CONFIG_K230_RPTUN_IPI */
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+int k230_rptun_init(const char *shmname, const char *cpuname)
+{
+  struct k230_rptun_dev_s  *dev = &g_rptun_dev;
+  int                       ret = OK;
+
+#ifdef CONFIG_K230_RPTUN_MASTER
+  /* master is responsible for initializing shmem */
+
+  memset((void *)SHMEM, 0, SHMEM_SIZE);
+  rpinfo("\ncleared %d @ %p\n", SHMEM_SIZE, SHMEM);
+  dev->master = true;
+#else
+  dev->master = false;
+#endif
+
+#ifdef CONFIG_K230_RPTUN_IPI
+  k230_ipi_init();
+#ifdef CONFIG_K230_RPTUN_MASTER
+  k230_rptun_ipi_master(dev);
+#else
+  k230_rptun_ipi_remote(dev);
+#endif
+#endif /* CONFIG_K230_RPTUN_IPI */
+
+  /* Configure device */
+
+  dev->rptun.ops = &g_k230_rptun_ops;
+  strncpy(dev->cpuname, cpuname, RPMSG_NAME_SIZE);
+  strncpy(dev->shmname, shmname, RPMSG_NAME_SIZE);
+
+  ret = rptun_initialize(&dev->rptun);
+  if (ret < 0)
+    {
+      rperr("ERROR: rptun_initialize failed %d!\n", ret);
+      goto errout;
+    }
+
+#ifdef CONFIG_K230_RPTUN_IPI
+  /* Create rptun RX thread */
+
+  ret = kthread_create("k230-rptun", CONFIG_RPTUN_PRIORITY,
+                       CONFIG_RPTUN_STACKSIZE, k230_rptun_thread, NULL);
```

Comment:
where is k230_rptun_thread defined?

Reply:
IPI support  is not available now, thus no need for that thread.

Reply:
IPI support  is unavailable yet, thus no need for that thread.